accepted/00083-containerized-proxy.md

```diff
@@ -0,0 +1,87 @@
+
+## Onboarding
+  - Containerized Proxy is not a managed client, does not run `salt-minion`, is not registered in the usual way
+  - A new UI page (and API endpoint) allows to add containerized Proxies by specifying their FQDN
+    - After confirmation a System entry (in the System list) is created 
+      - Normal client actions (package management, configuration file management...) will not be available on that System
+      - Proxy-specific Server UI/API functionality, like showing the list of attached clients, will be available like on regular Proxies
+    - for each such System, the Server generates a tarball ("configuration directory") that can be downloaded
+    - Proxy container needs the configuration directory to be mounted as a volume. After start, it is fully connected and operational
+
+## Operating
+  - Container will expose Salt (4505/4506) and https (443) ports
+  - Container will mount volumes:
+    - one with the configuration directory
+    - one with cached content (`/var/cache/squid/`, `/var/cache/rhn`)
+    - one with log files (`/var/log/apache2`, `/var/log/squid`, `salt-broker` log)
+  - Proxy Containers can be started/stopped/exchanged - as long as the configuration directory stays the same the Proxy identity, history and functionality will be preserved. Optionally, the cached content directory can also be preserved for optimal performance
+  - Creating bootstrap scripts from containerized Proxies will not be possible. `mgr-bootstrap --hostname=$PROXY_NAME` from the Server will still be possible
+  - SSL certificates can be handled in three ways:
+    1. using the same CA as the Server, as generated at Server installation time. A new SSL certificate for the Proxy is generated at "configuration tarball" creation time. This will be the default option
+    2. using the same CA as the Server, as generated by the user ("bring your own SSL certs"). A new SSL certificate for the Proxy has to be placed in the configuration directory by the user
+    3. leave SSL termination to the user's infrastructure altogether
+
+## Limitations
+  - Only ZeroMQ Salt clients are supported
+  - No monitoring functionality
+  - No SLP functionality
+  - Traceback emails are not sent
+  - `rhn_package_manager` is not present nor supported
+
+## Later steps
+  - add Salt SSH support
+  - enable the use of "configuration directories" for VM Proxies as well
+  - Retail Branch Server: to be defined in a separate RFC
+
+## Other details
+  - Container image includes Apache, Proxy-specific Python WSGI applications and `salt-broker`
+  - base image will be Leap 15.3
+  - API/UI will need the following data to add a Proxy:
+    - FQDN
+    - http proxy hostname, username and password (optional, default from [ProxySettingsManager.getProxySettings()](https://github.com/uyuni-project/uyuni/blob/b7cf9270cc1c5edd9c15fa1217e280d645bfcef9/java/code/src/com/redhat/rhn/manager/setup/ProxySettingsManager.java#L42))
```

Comment:
This is a bad idea according to my knowledge. The reason for this is that the hostname of a proxy is not stable.

- Docker: https://docs.docker.com/config/containers/container-networking/#ip-address-and-hostname
- K8s: https://kubernetes.io/docs/concepts/containers/container-environment/#container-information
- Podman: https://docs.podman.io/en/latest/markdown/podman-run.1.html#hostname-name-h

At least in the Docker usecase this means that the hostname is subject to change on image updates (and thus in a fat image package updates which happen regularly I hope).

In Kubernetes it could be that because of resource constrains or node drains the container is moved and then gets a new deployment id and thus new name.

We would need a proxy in front of our proxy who represents a stable entry point so that traffic (thus also the registration and communication of the proxy <-> server) is done in a reproducible manner.